File: Contests/AtCoderPython/AHC013/sample/sample_submission.py
import random
import sys
from copy import deepcopy

random.seed(1)


class Result:

    def __init__(self, moves, connects):
        self.moves = moves
        self.connects = connects


class Solver:
    USED = 9
    VECS = [
        [0, 1],
        [0, -1],
        [1, 0],
        [-1, 0],
    ]

    def __init__(self, N, K, C):
        self.N = N
        self.K = K
        self.C = C
        self.LIM = K * 100

    def _move(self, lim=None):
        if lim is None:
            lim = self.K * 50
        moves = []
        for _ in range(lim):
            i = random.randint(0, self.N - 1)
            j = random.randint(0, self.N - 1)
            if self.C[i][j] == 0:
                continue
            v = random.choice(self.VECS)
            i2 = i + v[0]
            j2 = j + v[1]
            if i2 < 0 or i2 >= self.N or j2 < 0 or j2 >= self.N:
                continue
            if self.C[i2][j2] != 0:
                continue
            self.C[i2][j2] = self.C[i][j]
            self.C[i][j] = 0
            moves.append([i, j, i2, j2])
        return moves

    def _connect(self, lim: int):
        connects = []

        def can_connect(i, j, v):
            i2 = i + v[0]
            j2 = j + v[1]
            while i2 < self.N and j2 < self.N:
                if self.C[i2][j2] == 0:
                    i2 += v[0]
                    j2 += v[1]
                    continue
                elif self.C[i2][j2] == self.C[i][j]:
                    return True
                return False
            return False

        def do_connect(i, j, v):
            i2 = i + v[0]
            j2 = j + v[1]
            while i2 < self.N and j2 < self.N:
                if self.C[i2][j2] == 0:
                    self.C[i2][j2] = self.USED
                elif self.C[i2][j2] == self.C[i][j]:
                    connects.append([i, j, i2, j2])
                    return
                else:
                    raise AssertionError()
                i2 += v[0]
                j2 += v[1]

        for i in range(self.N):
            for j in range(self.N):
                if self.C[i][j] in (0, self.USED):
                    continue
                for v in [[0, 1], [1, 0]]:
                    if can_connect(i, j, v):
                        do_connect(i, j, v)
                        if len(connects) >= lim:
                            return connects
        return connects

    def solve(self):
        # create random moves
        moves = self._move()
        # from each computer, connect to right and/or bottom if it will reach the same type
        connects = self._connect(self.LIM - len(moves))

        return Result(moves, connects)


class UnionFind:

    def __init__(self):
        self.parents = {}

    def find(self, x):
        if x not in self.parents:
            self.parents[x] = x
            return x
        elif self.parents[x] == x:
            return x
        else:
            self.parents[x] = self.find(self.parents[x])
            return self.parents[x]

    def unite(self, x, y):
        x = self.find(x)
        y = self.find(y)

        if x != y:
            self.parents[x] = y


def calc_score(N, K, C, res: Result):
    for v in res.moves:
        i, j, i2, j2 = v
        assert 1 <= C[i][j] <= K
        assert C[i2][j2] == 0
        C[i2][j2] = C[i][j]
        C[i][j] = 0

    uf = UnionFind()
    for v in res.connects:
        i, j, i2, j2 = v
        p1 = (i, j)
        p2 = (i2, j2)
        assert 1 <= C[i][j] <= K
        assert 1 <= C[i2][j2] <= K
        uf.unite(p1, p2)

    computers = []
    for i in range(N):
        for j in range(N):
            if 1 <= C[i][j] <= K:
                computers.append((i, j))

    score = 0
    for i in range(len(computers)):
        for j in range(i + 1, len(computers)):
            c1 = computers[i]
            c2 = computers[j]
            if uf.find(c1) != uf.find(c2):
                continue

            if C[c1[0]][c1[1]] == C[c2[0]][c2[1]]:
                score += 1
            else:
                score -= 1

    return max(score, 0)


def print_answer(res: Result):
    print(len(res.moves))
    for arr in res.moves:
        print(*arr)
    print(len(res.connects))
    for arr in res.connects:
        print(*arr)


def main():
    N, K = [int(v) for v in input().split(" ")]
    C = []

    for _ in range(N):
        C.append([int(v) for v in input()])

    solver = Solver(N, K, deepcopy(C))
    res = solver.solve()
    print(f"Score = {calc_score(N, K, deepcopy(C), res)}", file=sys.stderr)

    print_answer(res)


if __name__ == "__main__":
    main()

File: Contests/AtCoderPython/AHC013/sample/sample_submission.cpp
#include <cassert>
#include <iostream>
#include <map>
#include <random>
#include <utility>
#include <vector>

using namespace std;


struct MoveAction {
    int before_row, before_col, after_row, after_col;
    MoveAction(int before_row, int before_col, int after_row, int after_col) : 
        before_row(before_row), before_col(before_col), after_row(after_row), after_col(after_col) {}
};

struct ConnectAction {
    int c1_row, c1_col, c2_row, c2_col;
    ConnectAction(int c1_row, int c1_col, int c2_row, int c2_col) : 
        c1_row(c1_row), c1_col(c1_col), c2_row(c2_row), c2_col(c2_col) {}
};

struct Result {
    vector<MoveAction> move;
    vector<ConnectAction> connect;
    Result(const vector<MoveAction> &move, const vector<ConnectAction> &con) : move(move), connect(con) {}
};

struct Solver {
    static constexpr char USED = 'x';
    static constexpr int DR[4] = {0, 1, 0, -1};
    static constexpr int DC[4] = {1, 0, -1, 0};

    int N, K;
    int action_count_limit;
    mt19937 engine;
    vector<string> field;

    Solver(int N, int K, const vector<string> &field, int seed = 0) : 
        N(N), K(K), action_count_limit(K * 100), field(field)
    {
        engine.seed(seed);
    }

    bool can_move(int row, int col, int dir) const
    {
        int nrow = row + DR[dir];
        int ncol = col + DC[dir];
        if (0 <= nrow && nrow < N && 0 <= ncol && ncol < N) {
            return field[nrow][ncol] == '0';
        }
        return false;
    }

    vector<MoveAction> move(int move_limit = -1)
    {
        vector<MoveAction> ret;
        if (move_limit == -1) {
            move_limit = K * 50;
        }

        for (int i = 0; i < move_limit; i++) {
            int row = engine() % N;
            int col = engine() % N;
            int dir = engine() % 4;
            if (field[row][col] != '0' && can_move(row, col, dir)) {
                swap(field[row][col], field[row + DR[dir]][col + DC[dir]]);
                ret.emplace_back(row, col, row + DR[dir], col + DC[dir]);
                action_count_limit--;
            }
        }

        return ret;
    }

    bool can_connect(int row, int col, int dir) const
    {
        int nrow = row + DR[dir];
        int ncol = col + DC[dir];
        while (0 <= nrow && nrow < N && 0 <= ncol && ncol < N) {
            if (field[nrow][ncol] == field[row][col]) {
                return true;
            } else if (field[nrow][ncol] != '0') {
                return false;
            }
            nrow += DR[dir];
            ncol += DC[dir];
        }
        return false;
    }

    ConnectAction line_fill(int row, int col, int dir)
    {
        int nrow = row + DR[dir];
        int ncol = col + DC[dir];
        while (0 <= nrow && nrow < N && 0 <= ncol && ncol < N) {
            if (field[nrow][ncol] == field[row][col]) {
                return ConnectAction(row, col, nrow, ncol);
            }
            assert(field[nrow][ncol] == '0');
            field[nrow][ncol] = USED;
            nrow += DR[dir];
            ncol += DC[dir];
        }
        assert(false);
    }

    vector<ConnectAction> connect()
    {
        vector<ConnectAction> ret;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (field[i][j] != '0' && field[i][j] != 'x') {
                    for (int dir = 0; dir < 2; dir++) {
                        if (can_connect(i, j, dir)) {
                            ret.push_back(line_fill(i, j, dir));
                            action_count_limit--;
                            if (action_count_limit <= 0) {
                                return ret;
                            }
                        }
                    }
                }
            }
        }
        return ret;
    }

    Result solve()
    {
        // create random moves
        auto moves = move();
        // from each computer, connect to right and/or bottom if it will reach the same type
        auto connects = connect();
        return Result(moves, connects);
    }
};

struct UnionFind {
    map<pair<int,int>, pair<int, int>> parent;
    UnionFind() :parent() {}

    pair<int, int> find(pair<int, int> x)
    {
        if (parent.find(x) == parent.end()) {
            parent[x] = x;
            return x;
        } else if (parent[x] == x) {
            return x;
        } else {
            parent[x] = find(parent[x]);
            return parent[x];
        }
    }

    void unite(pair<int, int> x, pair<int, int> y)
    {
        x = find(x);
        y = find(y);
        if (x != y) {
            parent[x] = y;
        }
    }
};

int calc_score(int N, vector<string> field, const Result &res)
{
    for (auto r : res.move) {
        assert(field[r.before_row][r.before_col] != '0');
        assert(field[r.after_row][r.after_col] == '0');
        swap(field[r.before_row][r.before_col], field[r.after_row][r.after_col]);
    }

    UnionFind uf;
    for (auto r : res.connect) {
        pair<int, int> p1(r.c1_row, r.c1_col), p2(r.c2_row, r.c2_col);
        uf.unite(p1, p2);
    }

    vector<pair<int, int>> computers;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (field[i][j] != '0') {
                computers.emplace_back(i, j);
            }
        }
    }

    int score = 0;
    for (int i = 0; i < (int)computers.size(); i++) {
        for (int j = i+1; j < (int)computers.size(); j++) {
            auto c1 = computers[i];
            auto c2 = computers[j];
            if (uf.find(c1) == uf.find(c2)) {
                score += (field[c1.first][c1.second] == field[c2.first][c2.second]) ? 1 : -1;
            }
        }
    }

    return max(score, 0);
}

void print_answer(const Result &res)
{
    cout << res.move.size() << endl;
    for (auto m : res.move) {
        cout << m.before_row << " " << m.before_col << " "
            << m.after_row << " " << m.after_col << endl;
    }
    cout << res.connect.size() << endl;
    for (auto m : res.connect) {
        cout << m.c1_row << " " << m.c1_col << " "
            << m.c2_row << " " << m.c2_col << endl;
    }
}


int main()
{
    int N, K;
    cin >> N >> K;
    vector<string> field(N);
    for (int i = 0; i < N; i++) {
        cin >> field[i];
    }

    Solver s(N, K, field);
    auto ret = s.solve();

    cerr << "Score = " << calc_score(N, field, ret) << endl;

    print_answer(ret);

    return 0;
}
